### ISSUE TITLE
[Bug] Qwen3-TTS same-seed output diverges on first use of an exact prompt at concurrency=1

### ISSUE THREAD

**Issue by yuangxu** [sgl-project/sglang-omni#1428]
<html><head></head><body><h1>[Bug/Discussion] Radix cache prefix hit changes FlashInfer FA2 kernel specialization, producing non-deterministic TTS output (cache-invariance gap)</h1>
<h2>TL;DR</h2>
<p>With identical prompt, identical reference wav, identical seed and identical sampling params, Qwen3-TTS on SGLang-Omni produces <strong>two different waveforms</strong> depending on whether the request hits the radix cache.</p>
<p>The root cause is <strong>not</strong> KV corruption in the radix cache. The cached prefix KV is bitwise equal to the recomputed prefix KV. The actual chain is:</p>
<pre><code>radix prefix hit
  → q_len changes (58 → 1)
  → FlashInfer FA2 selects a different CTA_TILE_Q specialization
  → different floating-point reduction order
  → ~1 bf16 ULP-scale difference in attention output
  → top-k reordering in the talker logits
  → same seed, same Gumbel draw, same sampled rank, but a different token at that rank
  → autoregressive divergence
  → different wav
</code></pre>
<p>I'd like to discuss (a) whether this should be scoped as an SGLang issue, a FlashInfer issue, or an omni-path issue, and (b) which of the candidate fixes below is worth turning into a PR.</p>
<hr>
<h2>Environment</h2>
<ul>
<li>SGLang-Omni</li>
<li>Qwen3-TTS-12Hz-0.6B-Base</li>
<li>Single RTX 4090 (SM89), bf16</li>
<li>Attention backend: FlashInfer (FA2)</li>
</ul>
<hr>
<h2>Reproduction</h2>
<p>Send the exact same request repeatedly (same text, same reference wav, same seed, same sampling params).</p>
<p><strong>Radix cache ON:</strong></p>
<pre><code>Request 1 → A
Request 2 → B
Request 3 → B
Request 4 → B
Request 5 → B
</code></pre>
<p>The <code>ABBBB</code> pattern is stable across repeated rounds. Flushing the cache resets it to <code>A</code> again.</p>
<p><strong>Radix cache OFF:</strong></p>
<pre><code>AAAAA
AAAAA
AAAAA
</code></pre>
<p>So the mapping is exactly:</p>
<pre><code>radix MISS → A
radix HIT  → B
</code></pre>
<p>This is not jitter — it is fully reproducible and deterministic <em>given the cache state</em>.</p>
<hr>
<h2>What changes between MISS and HIT</h2>

  | cached tokens | new tokens | q_len | kv_len
-- | -- | -- | -- | --
MISS | 0 | 58 | 58 | 58
HIT | 57 | 1 | 1 | 58

<p>There is a second specialization switch further up (around <code>q_len=33</code>, CTA64 → CTA128), but that one is <strong>bitwise identical</strong> (A → A). So the claim is <em>not</em> "any kernel switch changes the result" — it is specifically that <strong>CTA16 is the divergent tier</strong> relative to CTA64/128.</p>
<p>A supplementary sweep on the KV side: <code>kv_len &lt;= 16</code> gives identical results across shapes, <code>kv_len &gt;= 17</code> starts to diverge.</p>
<h3>3. Different <code>q_len</code>, forced same <code>KernelTraits</code> — the causal experiment</h3>
<p>Kept <code>q_len = 1</code> vs <code>q_len = 58</code>, but forced both to use the same <code>KernelTraits</code> / same <code>CTA_TILE_Q</code> / same specialization, then compared the attention output for the last token:</p>
<pre><code>Attention O → bitwise identical
LSE         → bitwise identical
</code></pre>
<p>This nails the causality:</p>
<blockquote>
<p><code>q_len</code> is the dispatch trigger, not the numerical root cause. The numerical difference comes from the different floating-point reduction order between FA2 specializations.</p>
</blockquote>
<hr>
<h2>How a ~1 ULP difference becomes a different waveform</h2>
<p>Bit-level diff of the last token's attention output (16 × 128 = 2048 elements):</p>
<ul>
<li>1896 elements equal, 152 different</li>
<li>max absolute error ≈ 2^-12</li>
<li>max ULP distance = 121 (ULP distance varies with exponent range, so a small absolute error does not imply 1 ULP everywhere — but this is still a very small perturbation)</li>
</ul>
<p>Following it forward, the first discrete divergence is at:</p>
<pre><code>generation step   = 0
residual codebook = 3

MISS → token 68
HIT  → token 482
</code></pre>
<p>And here is the part I think is independently interesting:</p>
<ul>
<li>seed: identical</li>
<li>Gumbel noise draw: identical</li>
<li>sampled <strong>rank</strong>: identical — both picked <code>rank = 27</code></li>
</ul>
<p>But <code>rank 27</code> maps to token 68 on one side and token 482 on the other, because the tiny logits perturbation reordered the top-50. Roughly <strong>37 of 50</strong> positions changed relative order.</p>
<p>So <strong>Gumbel noise is indexed by rank position rather than by token ID.</strong> That means any perturbation that reorders the candidate set changes the sampled token even when the RNG stream is perfectly reproduced. In a text LLM this would usually be invisible. In TTS it selects a different codec token at step 0, which changes the next input, which diverges every subsequent codebook — and the result is an audibly different wav.</p>
<p>This is the severity gap: <strong>TTS amplifies a 1-ULP attention difference into a fully different output</strong>, in a way text generation typically does not.</p>
<hr>
<h2>Why I'd call this a <em>cache-invariance</em> gap, not just a radix bug</h2>
<p>SGLang already reasons carefully about batch invariance — same request, different batch composition, same result. What this case shows is a sibling property that isn't currently guaranteed:</p>
<blockquote>
<p>Same logical request, different <strong>execution shape</strong> (prefix cache state, chunk boundary, padding, scheduler decision), same result.</p>
</blockquote>
<p>Prefix caching is one way execution shape changes; batching is another. The failure mode is the same:</p>
<pre><code>same logical request
  → different execution shape
  → different kernel specialization
  → different numerical output
</code></pre>
<h2>What <code>--enable-deterministic-inference</code> currently does</h2>
<p>On the version I tested, <code>--enable-deterministic-inference</code> with <code>--attention-backend flashinfer</code> avoids the FlashInfer + radix cache combination altogether. Every request goes through full prefill:</p>
<pre><code>cached = 0, new = 58  → q_len = 58, every time → always A
</code></pre>
<p>That is stable, but it works by <strong>preventing the execution shape from ever varying</strong>, not by making the kernels shape-invariant. It treats the trigger rather than the root cause, and it costs the entire prefix-cache benefit. It also reads to me more like a backend deterministic-coverage gap than a Qwen3-TTS-specific problem.</p>
<hr>
<h2>Candidate directions (roughly increasing cost)</h2>
<p><strong>A. Clamp prefix match length so <code>extend_len &gt;= 9</code>.</strong> Cheapest and most self-contained: never let a prefix hit drive <code>q_len</code> into the CTA16 tier. Costs a handful of recomputed tokens, keeps most of the cache benefit. Probably the most mergeable as a first PR.</p>
<p><strong>B. Lock the tile size in deterministic mode inside FlashInfer.</strong> Pin <code>CTA_TILE_Q</code> when deterministic inference is on, so dispatch stops being shape-dependent. Cleaner semantically, needs FlashInfer-side changes.</p>
<p><strong>C. Canonical KV partition scheme.</strong> Decouple partition boundaries from tile/warp layout so the reduction order is a function of the logical attention rather than the launch shape. This is a FlashInfer design proposal, not a quick fix, but it's the only one that makes the property structural.</p>
<p><strong>D. Index Gumbel noise by token ID instead of rank.</strong> Independent of the attention issue. Doesn't make attention invariant, but removes the amplifier that turns a reordering into a different token. Worth considering on its own merits for the omni sampling path.</p>
<hr>
<h2>Open questions for maintainers</h2>
<ol>
<li><strong>Scoping</strong> — should this be filed primarily against SGLang (dispatch/scheduling), FlashInfer (kernel invariance), or the omni sampling path? My current view is that no single component is wrong in isolation, which makes attribution genuinely ambiguous.</li>
<li>Is the <code>q_len = 8/9</code> threshold <strong>model-architecture dependent</strong> (e.g. via <code>gqa_group_size</code>)? If so, option A needs the bound computed rather than hardcoded.</li>
<li>Does <strong>chunked prefill</strong> hit the same condition when a chunk boundary lands such that the final chunk has <code>q_len &lt;= 8</code>? I haven't tested this, but it looks structurally identical.</li>
<li>Is there an existing upstream FlashInfer issue tracking CTA16 vs CTA64/128 numerical divergence? I couldn't find one, but I'd rather not duplicate.</li>
<li>Is cache invariance something SGLang wants to state as an explicit guarantee alongside batch invariance, or is "deterministic mode disables prefix cache" the intended contract?</li>
</ol>
<p>I'm happy to write up A as a PR, or to run additional sweeps (other models, other <code>kv_len</code> ranges, chunked prefill) if that would help the discussion. Repro scripts for the isolated FlashInfer replay and the <code>q_len</code> sweep are available on request.</p></body></html>

**Comment by lzwnoname** [comment]
Hi, I'd like to take this issue. Could you please assign it to me? Thanks!

**Comment by yuangxu** [comment]
@lzwnoname Thanks for the interest! I'm already actively investigating this issue and would like to continue working on it for now.
I'm running the Radix Cache ablation next, followed by CUDA Graph / determinism-related checks, and I'll post the results here as they come.
If you'd like to help, happy to coordinate on a separate subtest. Thanks!

**Comment by lzwnoname** [comment]
@yuangxu I'm very happy to help. How can I join this work?
